bundler/lib/bundler/match_platform.rb

```diff
@@ -11,11 +11,16 @@ def match_platform(p)
     end
 
     def self.platforms_match?(gemspec_platform, local_platform)
+      # p({ gem: gemspec_platform, local: local_platform })
       return true if gemspec_platform.nil?
-      return true if Gem::Platform::RUBY == gemspec_platform
-      return true if local_platform == gemspec_platform
+      return true if gemspec_platform == Gem::Platform::RUBY
+      return true if gemspec_platform == local_platform
       gemspec_platform = Gem::Platform.new(gemspec_platform)
-      return true if GemHelpers.generic(gemspec_platform) === local_platform
+      #return true if GemHelpers.generic(gemspec_platform) === local_platform
```

Comment:
That's the second question I had. I added a test to make sure that without a `ruby` nor a `linux` gem, a `linux-musl` gem by itself would never get picked by a `linux` runtime, but the resolver still returned the `linux-musl` gem!

It was really unclear why, so I went through the whole resolver thing to find that it's not the `linux` platform `spec_group` but the additional `spec_group_ruby` that causes this, because:

```
pry> Bundler::GemHelpers.generic(Gem::Platform.new('x86_64-linux-musl'))
=> "ruby"
```

Which equals `local_platform from the resolver's `spec_group_ruby` in `search_for`.

This means that in the results, `linux-musl` still appears for `linux` through the added `ruby` candidate spec group, and since `specificity_score` doesn't remove it, it bubbles up to the final result even though it's absent from the platform-specific `spec_group`...

The possible leakage also applies in other cases (such as `linux` appearing for `linux-musl` through `ruby`) but the end result is so that this is equivalent to what we want (and what appears in the `spec_group`). If I got this right, I suspect in most cases the problem is hidden this way but it could cause subtle issues here and there.

Reply:
Some platforms are "forced" to not have a generic platform (i.e., only match their own platform).

https://github.com/rubygems/rubygems/blob/5cfad4bd6661f2370bff0aaef1e81802420d2cef/bundler/lib/bundler/gem_helpers.rb#L6-L14

Perhaps our constraints for `x86_64-linux-musl` match this case.

Reply:
@lloeki This line has now been removed, I think it was indeed incorrect. So this should no longer be an issue!